Sources/AsyncCollections/Filter.swift

```diff
@@ -0,0 +1,123 @@
+
+extension Sequence where Element: Sendable {
+    /// Returns an array containing, in order, the elements of the sequence
+    /// that satisfy the given predicate.
+    ///
+    /// - Parameter isIncluded: An async closure that takes an element of the
+    ///   sequence as its argument and returns a Boolean value indicating
+    ///   whether the element should be included in the returned array.
+    /// - Returns: An array of the elements that `isIncluded` allowed.
+    public func asyncFilter(_ isIncluded: @Sendable (Element) async throws -> Bool) async rethrows -> [Element] {
+        var result = ContiguousArray<Element>()
+
+        var iterator = self.makeIterator()
+
+        while let element = iterator.next() {
+            if try await isIncluded(element) {
+                result.append(element)
+            }
+        }
+
+        return Array(result)
+    }
+
+    /// Returns an array containing, in order, the elements of the sequence
+    /// that satisfy the given predicate.
+    ///
+    /// This differs from `asyncFilter` in that it uses a `TaskGroup` to run the transform
+    /// closure for all the elements of the Sequence. This allows all the transform closures
+    /// to run concurrently instead of serially. Returns only when the closure has been run
+    /// on all the elements of the Sequence.
+    /// - Parameters:
+    ///   - priority: Task priority for tasks in TaskGroup
+    ///   - isIncluded: An async closure that takes an element of the
+    ///   sequence as its argument and returns a Boolean value indicating
+    ///   whether the element should be included in the returned array.
+    /// - Returns: An array of the elements that `isIncluded` allowed.
+    public func concurrentFilter(priority: TaskPriority? = nil, _ isIncluded: @Sendable @escaping (Element) async throws -> Bool) async rethrows -> [Element] {
+        let result: ContiguousArray<(Int, Element)> = try await withThrowingTaskGroup(
+            of: (Int, Element)?.self
+        ) { group in
+            for (index, element) in self.enumerated() {
+                group.addTask(priority: priority) {
+                    if try await isIncluded(element) {
+                        return (index, element)
+                    } else {
+                        return nil
+                    }
+                }
+            }
+            // Code for collating results copied from Sequence.map in Swift codebase
+            let initialCapacity = underestimatedCount
```

Comment:
Don't think you should use `underestimatedCount` here. That is based off the size of the original sequence. No point allocating space for the original sequence if you are only returning one value.

Reply:
Right, i thought maybe it's still worth it to pre-allocate even for a filter ...
I know stuff like std-lib implementation don't like to do these kinds of assumptions so i guess i'll just remove that.